### PR TITLE
ci: add MegaLinter and fix its output

### PR DESCRIPTION
**Pull request by nvuillam**
- Setup MegaLinter job in global PR workflow
- Fix or ignore issues detected by MegaLinter (mostly formatting, dead links and typos)
- Manage concurrency for PR job

### CODE REVIEW

**Comment by nvuillam** [comment]
@scolladon I've been careful for the few js code I updated... I don't see what could trigger such error :/

https://github.com/scolladon/sfdx-git-delta/actions/runs/4708313450/jobs/8350686243

Any idea ? :)

**Comment by scolladon** [comment]
> @scolladon I've been careful for the few js code I updated... I don't see what could trigger such error :/
> 
> [scolladon/sfdx-git-delta/actions/runs/4708313450/jobs/8350686243](https://github.com/scolladon/sfdx-git-delta/actions/runs/4708313450/jobs/8350686243)
> 
> Any idea ? :)

That is expected, this is related to the next way to deal with particular metadata file.
Current main does not comply with this new way to deal with particular metadata file but e2e test are already prepared to support this feature.

File: .github/linters/.cspell.json [hunk 1]
```diff
@@ -0,0 +1,146 @@
+{
+    "ignorePaths": [
+        "**/node_modules/**",
+        "**/vscode-extension/**",
+        "**/.git/**",
+        "**/.pnpm-lock.json",
+        ".vscode",
+        "megalinter",
+        "package-lock.json",
+        "report"
```
Comment (scolladon) [inline .github/linters/.cspell.json:10]: **suggestion:** maybe couples lines are not needed here
and we could add `yarn.lock` and `CHANGELOG.md`?

```suggestion
	"CHANGELOG.md",
        "**/node_modules/**",
        "**/.git/**",
        ".vscode",
        "megalinter",
        "yarn.lock",
        "report"
```

File: .github/linters/.jscpd.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    "**/node_modules/**",
+    "**/.git/**",
+    "**/.rbenv/**",
+    "**/.venv/**",
+    "**/*cache*/**",
+    "**/.github/**",
+    "**/.idea/**",
+    "**/report/**",
+    "**/*.svg",
+    "**/__tests__/**"
```
Comment (scolladon) [inline .github/linters/.jscpd.json:14]:
```suggestion
    "**/node_modules/**",
    "**/.git/**",
    "**/*cache*/**",
    "**/.github/**",
    "**/report/**",
    "**/img/**",
    "**/__tests__/**"
```

File: .github/workflows/on-main-push.yml [hunk 3]
```diff
@@ -27,10 +28,10 @@ jobs:
           package-name: sfdx-git-delta
           token: ${{ secrets.RELEASE_PAT }}
 
-  release: 
+  release:
     needs: prepare-release
     runs-on: ubuntu-latest
-    if: ${{ needs.prepare-release.outputs.release_created == 'true' }}
+    if: needs.prepare-release.outputs.release_created == 'true'
```
Comment (scolladon) [inline .github/workflows/on-main-push.yml:34]: **thought:** I'm curious to see if it works

File: .github/workflows/on-pull-request.yml [hunk 4]
```diff
@@ -76,6 +81,40 @@ jobs:
...
+      # Upload MegaLinter artifacts
+      - name: Archive production artifacts
+        if: ${{ success() }} || ${{ failure() }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: MegaLinter reports
+          path: |
+            megalinter-reports
+            mega-linter.log
+
```
Comment (scolladon) [inline .github/workflows/on-pull-request.yml:117]:
```suggestion
     - uses: actions/upload-artifact@v3
        with:
          name: megalinter-reports
          path: |
            megalinter-reports
            mega-linter.log

```

File: CHANGELOG.md
Comment (scolladon) [inline CHANGELOG.md:1]: **suggestion:** it could be ignored, this file is auto generated

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -215,11 +215,12 @@ _note: If your pull request needs more changes, keep working on your feature bra
...
-[configuring-a-remote-for-a-fork]: https://help.github.com/en/articles/configuring-a-remote-for-a-fork
-[setup-github-ssh]: https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/
-[creating-a-pull-request]: https://help.github.com/articles/creating-a-pull-request/
-[eslint-integrations]: http://eslint.org/docs/user-guide/integrations
+- [fork-a-repo](https://help.github.com/en/articles/fork-a-repo)
+- [configuring-a-remote-for-a-fork](https://help.github.com/en/articles/configuring-a-remote-for-a-fork)
+- [setup-github-ssh](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/)
+- [creating-a-pull-request](https://help.github.com/articles/creating-a-pull-request/)
+- [eslint-integrations](http://eslint.org/docs/user-guide/integrations)
+- [MegaLinter](https://megalinter.io)
```
Comment (scolladon) [inline CONTRIBUTING.md:223]: **praises:** nice addition, thanks !

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -200,7 +200,7 @@ Test your change by running the unit tests and integration tests. Instructions [
 ### Create a pull request
 
 If you've never created a pull request before, follow [these
-instructions][creating-a-pull-request]. Pull request samples [here](https://github.com/salesforce/sfdx-git-delta/pulls)
+instructions][creating-a-pull-request]. Pull request samples [here](https://github.com/scolladon/sfdx-git-delta/pulls)
```
Comment (scolladon) [inline CONTRIBUTING.md:203]: **praises:** nice ! boy scout attitude rule 👍

File: __tests__/integration/delta.test.js [hunk 7]
```diff
@@ -29,10 +29,10 @@ const lines = [
...
   'A      force-app/main/default/classes/controllers/Added.cls-meta.xml',
   'D      force-app/main/default/classes/controllers/Deleted.cls',
-  'A      force-app/main/default/staticressources/Added.resource-meta.xml',
-  'A      force-app/main/default/staticressources/Added.zip',
-  'M      force-app/main/default/staticressources/Added/changed.png',
-  'D      force-app/main/default/staticressources/Added/deleted.png',
+  'A      force-app/main/default/staticresources/Added.resource-meta.xml',
+  'A      force-app/main/default/staticresources/Added.zip',
+  'M      force-app/main/default/staticresources/Added/changed.png',
+  'D      force-app/main/default/staticresources/Added/deleted.png',
```
Comment (scolladon) [inline __tests__/integration/delta.test.js:35]: **praises:** nice ! boy scout attitude rule, thanks

**Review by scolladon — COMMENTED** [review]
LGTM

Couple of suggestion we can discuss about
Very nice addition, thank you very much for this 🙏 ❤️ 
I look forward using `MegaLinter` !

File: .github/workflows/on-main-push.yml (hunk 3 above)
Comment (nvuillam) [reply .github/workflows/on-main-push.yml:34]: actionlint detected this one, and the doc confirms :)
image

File: .github/workflows/on-pull-request.yml (hunk 4 above)
Comment (nvuillam) [reply .github/workflows/on-pull-request.yml:117]: I'm pretty sure you need this one, else the default condition is `success()`
But you obviously want to upload reports when MegaLinter fails :)

Comment (scolladon) [reply .github/workflows/on-pull-request.yml:117]: Then we could write it this way?

if: always()

Comment (nvuillam) [reply .github/workflows/on-pull-request.yml:117]: I think it did not exist at the time MegaLinter was created, but it's worth a try :)
If it works I'll even change the default ML workflow ^^

File: .github/workflows/on-pull-request.yml [hunk 8]
```diff
@@ -8,6 +9,11 @@ on:
       - "**.md"
       - "img/**"
 
+# Manage concurrency to stop running jobs and start new ones in case of new commit pushed
+concurrency:
+  group: ${{ github.ref }}-${{ github.workflow }}
+  cancel-in-progress: true
+
```
Comment (scolladon) [inline .github/workflows/on-pull-request.yml:16]: **praises:** nice boy scout attitude here as well, thanks

**Review by scolladon — DISMISSED** [review]
LGTM

Great contribution, thanks @nvuillam !

**Comment by scolladon** [comment]
@nvuillam I think there is an issue with the upload of artifact.
Probably from my inline suggestion via github comment.
You may want to fix it locally on your end 

Screenshot 2023-04-16 at 11 26 11

**Comment by nvuillam** [comment]
> @nvuillam I think there is an indentation issue with the "upload of artifact" step. Probably from my inline suggestion via github comment. You may want to fix it locally on your end
> 
> Screenshot 2023-04-16 at 11 26 11

Done :)

**Review by scolladon — APPROVED** [review]
LGTM

@nvuillam any idea why the github bot does not write comment with the `megalinter` execution result?

**Comment by nvuillam** [comment]
@scolladon it's because the PR is from a fork
GITHUB_TOKEN from nvuillam/sfdx-git-delta does not have rights to write a comment on your repo, but it will work for PR coming from scolladon/sfdx-git-delta

**Comment by nvuillam** [comment]
There is a long time I wonder how to make that work... maybe it could be done if the comment was written by MegaLinter bot, but it probably would require some storage of MegaLinter PAT on a third party server... I'll let you know when it'll be possible :)

**Review by mehdisfdc — APPROVED** [review]

**Pull request merged**